File: packages/model/Cycles/TaskInstance.cs
using System.ComponentModel;
using System.Text.Json.Serialization;
using Anovase.Sunnyside.Backlog;

namespace Anovase.Sunnyside.Cycles;

public enum TaskStatus
{
	Open = 0,
	Done = 1,
	Delegated = 2,
	Failed = -1,
}

public class TaskInstance
{
	// Meta
	public Guid Id { get; set; }
	public Directive Directive { get; set; } = null!;

	[JsonIgnore]
	public Cycle Cycle { get; set; } = null!;
	public Guid CycleId { get; set; } = Guid.Empty;

	// Properties
	public string? Name { get; set; }
	public double TimeTracked { get; set; }
	public double TimeAllocated { get; set; }
	public TaskObjective? Objective { get; set; }
	public TaskStatus Status { get; set; } = TaskStatus.Open;
}

File: packages/client/web/sdk/models/TaskInstance.ts
import { Cycle, Directive, TaskObjective } from '.'

export const enum TaskStatus {
	open = 0,
	done = 1,
	delegated = 2,
	failed = -1,
}

export class TaskInstance {
	id = '00000000-0000-0000-0000-000000000000'
	cycleId = '00000000-0000-0000-0000-000000000000'
	directive: Directive = null! //TODO: can this be nullable?

	name?: string
	timeAllocated = 0
	timeTracked = 0
	objective? = TaskObjective.doneOnly
	status = TaskStatus.open

	get done() { return this.status === TaskStatus.done }
	set done(value: boolean) { this.status = value ? TaskStatus.done : TaskStatus.open }

	constructor(cycleId: string) {
		this.cycleId = cycleId
	}
}
